src/config.zig

```diff
@@ -1,413 +1,198 @@
-/// The default maximum amount of memory to use.
-pub const memory_size_max_default = 1024 * 1024 * 1024;
+pub const Config = struct {
+    pub const Cluster = ConfigCluster;
+    pub const Process = ConfigProcess;
 
-/// The maximum number of accounts to store in memory:
-/// This impacts the amount of memory allocated at initialization by the server.
-pub const cache_accounts_max = switch (deployment_environment) {
-    .production => 64 * 1024,
-    else => 8 * 1024,
+    cluster: ConfigCluster,
+    process: ConfigProcess,
 };
 
-/// The maximum number of transfers to store in memory:
-/// This impacts the amount of memory allocated at initialization by the server.
-/// We allocate more capacity than the number of transfers for a safe hash table load factor.
-pub const cache_transfers_max = switch (deployment_environment) {
-    .production => 1024 * 1024,
-    else => 64 * 1024,
-};
-
-/// The maximum number of two-phase transfers to store in memory:
-/// This impacts the amount of memory allocated at initialization by the server.
-pub const cache_transfers_posted_max = cache_transfers_max;
-
-/// The maximum number of batch entries in the journal file:
-/// A batch entry may contain many transfers, so this is not a limit on the number of transfers.
-/// We need this limit to allocate space for copies of batch headers at the start of the journal.
-/// These header copies enable us to disentangle corruption from crashes and recover accordingly.
-pub const journal_slot_count = switch (deployment_environment) {
-    .production => 1024,
-    else => 1024,
+/// Configurations which are tunable per-replica (or per-client).
+/// - Replica configs need not equal each other.
+/// - Client configs need not equal each other.
+/// - Client configs need not equal replica configs.
+/// - Replica configs can change between restarts.
+///
+/// Fields are documented within the top-level module body.
+const ConfigProcess = struct {
+    log_level: std.log.Level = .info,
+    tracer_backend: TracerBackend = .none,
+    verify: bool,
+    port: u16 = 3001,
+    address: []const u8 = "127.0.0.1",
+    memory_size_max_default: u64 = 1024 * 1024 * 1024,
+    cache_accounts_max: usize,
+    cache_transfers_max: usize,
+    cache_transfers_posted_max: usize,
+    client_request_queue_max: usize = 32,
+    lsm_manifest_node_size: usize = 16 * 1024,
+    connection_delay_min_ms: u64 = 50,
+    connection_delay_max_ms: u64 = 1000,
+    tcp_backlog: u31 = 64,
+    tcp_rcvbuf: c_int = 4 * 1024 * 1024,
+    tcp_keepalive: bool = true,
+    tcp_keepidle: c_int = 5,
+    tcp_keepintvl: c_int = 4,
+    tcp_keepcnt: c_int = 3,
+    tcp_nodelay: bool = true,
+    direct_io: bool,
+    direct_io_required: bool,
+    journal_iops_read_max: usize = 8,
+    journal_iops_write_max: usize = 8,
+    tick_ms: u63 = 10,
+    rtt_ms: u64 = 300,
+    rtt_multiple: u8 = 2,
+    backoff_min_ms: u64 = 100,
+    backoff_max_ms: u64 = 10000,
+    clock_offset_tolerance_max_ms: u64 = 10000,
```

Comment:
I haven't looked at how the clock sync works yet, but I'm suprised that different replicas can have different settings for it.

Reply:
You probably _shouldn't_ use different values, but we don't rely on clock sync for correctness, only availability. `ConfigProcess` is for things that are safe to change at "cluster runtime" — e.g. you could do a rolling replica restart to deploy a new value.

Reply:
:+1: